src/System.IO.FileSystem/src/System/IO/PathHelpers.Windows.cs

```diff
@@ -0,0 +1,69 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+
+namespace System.IO
+{
+    internal static partial class PathHelpers
+    {
+        /// <summary>
+        /// Return fully normalized path. It uses ArrayPool to reduce allocation.
+        /// </summary>
+        /// <param name="path"></param>
+        /// <param name="allowTrailingSeparator"></param>
+        internal unsafe static ReadOnlySpan<char> FastNormalizePath(string path, bool allowTrailingSeparator = false)
+        {
+            if (path.IndexOf('\0') != -1)
+                return ReadOnlySpan<char>.Empty;
+
+            if (PathInternal.IsExtended(path))
+            {
+                if (path.Length > 4 && path[path.Length - 1] == '\\' && path[path.Length - 2] != ':')
+                {
+                    return allowTrailingSeparator ? path.AsSpan().Slice(0, path.Length - 1) : ReadOnlySpan<char>.Empty;
+                }
+
+                return path.AsSpan();
+            }
+
+            uint result = 260;
+            char[] buffer = null;
+
+            do
+            {
+                if (buffer != null)
+                    ArrayPool<char>.Shared.Return(buffer);
+
+                buffer = ArrayPool<char>.Shared.Rent((int)result);
+
+                fixed (char* c = buffer)
+                {
+                    result = Interop.Kernel32.GetFullPathNameW(path, (uint)buffer.Length, c, IntPtr.Zero);
+                }
+
+                if (result == 0)
+                {
+                    return ReadOnlySpan<char>.Empty;
+                }
+            } while (result > buffer.Length);
+
+            if (buffer[result - 1] == '\\')
+            {
+                if (!allowTrailingSeparator)
+                {
+                    return ReadOnlySpan<char>.Empty;
+                }
+                else if (result != 3)
+                {
+                    // C:\ is a special case, we can't remove the trailing slash in \\?\ format.
+                    // There is no valid path that can come back at 3 characters other than C:\
+                    return new ReadOnlySpan<char>(buffer, 0, (int)result - 1);
+                }
+            }
+
+            return new ReadOnlySpan<char>(buffer, 0, (int)result);
```

Comment:
This is not a good use of ArrayPool. It's renting an array and then returning that array in a span... that array will never be returned to the pool. This is worse than just allocating an array, as it's taking out a likely long lived array that's stilled in a higher GC generation.

Reply:
```
    internal struct PooledBuffer : IDisposable
    {
        public char[] _buffer;

        public char[] Rent(int minLength)
        {
            if (_buffer != null)
                ArrayPool<char>.Shared.Return(_buffer);

            return _buffer = ArrayPool<char>.Shared.Rent(minLength);
        }


        public void Dispose()
        {
            if (_buffer != null)
                ArrayPool<char>.Shared.Return(_buffer);
            _buffer = null;
        }
    }
```
And using this struct to get the array ?.

Reply:
> And using this struct to get the array ?

I don't understand the question.  Can you elaborate?

Reply:
The problem with the current version is when we return, its not able to return the last rented array. So will using the Idisposable class to return the last rented array solves our leak problem ?

Reply:
using this struct like https://github.com/dotnet/corefx/compare/master...nchikanov:FileExistsSPAN#diff-2a5b1080a98c5e368dda4139a17c4915R174

Reply:
> using this struct like master...nchikanov:FileExistsSPANdiff-2a5b1080a98c5e368dda4139a17c4915R174

That link isn't working for me.

> So will using the Idisposable class to return the last rented array solves our leak problem ?

It depends on how it's consumed.

Reply:
@Anipik the primary problem is not returning it to the pool is that as @stephentoub points out ArrayPool is intended for temporary allocations -- in this case, assume that this same function that rents from the pool would also return to it before it completed. There's different ways to achieve that, but first you have to decide whether to use it and how. Right now you're pulling from the pool and giving it out like candy to the caller. The pool only works well if you hand it back.
@JeremyKuhne may be able to clarify what he had in mind.

Reply:
Okay I will talk it to Jeremy about it

Reply:
Thanks.  What's the status of this?  Should it be closed in the meantime?